keyboards/gmmk/numpad/keymaps/via/keymap.c

```diff
@@ -0,0 +1,90 @@
+/* Copyright 2021 Glorious, LLC <salman@pcgamingrace.com>
+Modified 2022 by rustedaperture for qmk_firmware
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include QMK_KEYBOARD_H
+#include "analog.h"
+#include "qmk_midi.h"
+
+// clang-format off
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+//      NUM      /       *       -
+//      7        8       9       +
+//      4        5       6       CALC
+//      1        2       3       RET
+//      0                        .
+
+  [0] = LAYOUT(
+    MO(1),   KC_PSLS,   KC_PAST,   KC_PMNS,
+    KC_P7,   KC_P8,     KC_P9,     KC_PPLS,
+    KC_P4,   KC_P5,     KC_P6,     KC_CALC,
+    KC_P1,   KC_P2,     KC_P3,     KC_PENT,
+    KC_P0,                         KC_PDOT
+  ),
+  [1] = LAYOUT(
+    _______,   KC_PSLS,   KC_PAST,   KC_PMNS,
+    KC_P7,     RGB_VAI,   KC_P9,     KC_PPLS,
+    RGB_RMOD,  KC_P5,     RGB_MOD,   KC_CALC,
+    KC_P1,     RGB_VAD,   KC_P3,     KC_PENT,
+    RGB_TOG,                         QK_BOOT
+  ),
+  [2] = LAYOUT(
+    _______,   _______,   _______,   _______,
+    _______,   _______,   _______,   _______,
+    _______,   _______,   _______,   _______,
+    _______,   _______,   _______,   _______,
+    _______,                         _______
+  ),
+  [3] = LAYOUT(
+    _______,   _______,   _______,   _______,
+    _______,   _______,   _______,   _______,
+    _______,   _______,   _______,   _______,
+    _______,   _______,   _______,   _______,
+    _______,                         _______
+  ),
+};
+// clang-format on
+
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
+    [0] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [1] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [2] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [3] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) }
+};
+
+// Potentiometer Slider, MIDI Control
+
+uint8_t divisor = 0;
+
+void slider(void) {
+    if (divisor++) { /* only run the slider function 1/256 times it's called */
+        return;
+    }
+    midi_send_cc(&midi_device, 2, 0x3E, 0x7F + (analogReadPin(SLIDER_PIN) >> 3));
+}
+
+extern int16_t enc;
+extern int16_t encPrev;
+
+void housekeeping_task_user(void) {
```

Comment:
wait, so you have the encoder wired up to a single pin, and then some manual behavior for the encoder push? 

If so, that's .... gross.  

While not necessarily better, having the encoder behavior in the normal matrix/layout would be better, IMO.  The work louder micro has a custom matrix with that sort of behavior, for reference.

Reply:
@drashna I will have to investigate once I get the chance. This was an implementation that I grabbed from [lynepad](https://github.com/qmk/qmk_firmware/tree/042e42d1b7de603bc0a164b7d3b2ec761c6e140f/keyboards/qvex/lynepad). I'm working with my limited knowledge and was grabbing from other keyboards as an example. 

As much as I would like to get it right from the start, I prefer to work by examples since I'm unfamiliar with the code. 

I do agree that it was a dirty fix though and I will attempt to get it working.

Reply:
for reference, the work louder micro has 2 touch sensors that pull a digital pin high or low, depending on the status.  If the board was set up as "direct pin" these could have been used directly.  But since they weren't... a custom matrix was needed. Mostly, I took the normal matrix, removed everything that wasn't needed, and then added this block: 

https://github.com/qmk/qmk_firmware/blob/8bb7d24fba6c70dd5865b4d92421fcb9248b7f0b/keyboards/work_louder/micro/matrix.c#L99-L106

So on row 3, at col 0 and 3, it will read the pin for the touch sensors rather than the normal matrix read attempt.  Basically, that's what you want.  And then the normal keymap/layout can be used and you don't need a hard coded behavior.

Reply:
@drashna I'm trying to get this to work for ROW2COL. I'm struggling to figure out where to put the check for row and col.

```
__attribute__((weak)) void matrix_read_rows_on_col(matrix_row_t current_matrix[], uint8_t current_col, matrix_row_t row_shifter) {    // Start with a clear matrix row
        bool key_pressed = false;

    // Select col
    if (!select_col(current_col)) { // select col
        return;                     // skip NO_PIN col
    }
    matrix_output_select_delay();

    // For each row...
    for (uint8_t row_index = 0; row_index < MATRIX_ROWS; row_index++) {
        // Check row pin state
        if (readMatrixPin(row_pins[row_index]) == 0) {
            // Pin LO, set col bit
            current_matrix[row_index] |= row_shifter;
            key_pressed = true;
        } else {
            // Pin HI, clear col bit
            current_matrix[row_index] &= ~row_shifter;
        }
    }

    // Unselect col
    unselect_col(current_col);
    matrix_output_unselect_delay(current_col, key_pressed); // wait for all Row signals to go HIGH
}
```

Reply:
I've set the `matrix_init_custom` to be the following 
```
__attribute__((weak)) void matrix_init_custom(void) {
    unselect_cols();
    for (uint8_t x = 0; x < MATRIX_ROWS; x++) {
        if (row_pins[x] != NO_PIN) {
            setPinInputHigh_atomic(row_pins[x]);
        }
    }
    setPinInputHigh_atomic(B8);
}
```

Reply:
Here is my `matrix.c` in its entirety 
```
// Copyright 2022 Cameron Varley (@RustedAperture)
// SPDX-License-Identifier: GPL-2.0-or-later

/*
 * scan matrix
 */
#include <stdint.h>
#include <stdbool.h>
#if defined(__AVR__)
#include <avr/io.h>
#endif
#include "wait.h"
#include "print.h"
#include "debug.h"
#include "util.h"
#include "matrix.h"
#include "debounce.h"
#include QMK_KEYBOARD_H

/* matrix state(1:on, 0:off) */
extern matrix_row_t matrix[MATRIX_ROWS];     // debounced values
extern matrix_row_t raw_matrix[MATRIX_ROWS]; // raw values

static const pin_t row_pins[MATRIX_ROWS] = MATRIX_ROW_PINS;
static const pin_t col_pins[MATRIX_COLS] = MATRIX_COL_PINS;
#define MATRIX_ROW_SHIFTER ((matrix_row_t)1)

static inline void setPinOutput_writeLow(pin_t pin) {
    ATOMIC_BLOCK_FORCEON {
        setPinOutput(pin);
        writePinLow(pin);
    }
}

static inline void setPinOutput_writeHigh(pin_t pin) {
    ATOMIC_BLOCK_FORCEON {
        setPinOutput(pin);
        writePinHigh(pin);
    }
}

static inline void setPinInputHigh_atomic(pin_t pin) {
    ATOMIC_BLOCK_FORCEON {
        setPinInputHigh(pin);
    }
}

static inline uint8_t readMatrixPin(pin_t pin) {
    if (pin != NO_PIN) {
        return readPin(pin);
    } else {
        return 1;
    }
}

static bool select_col(uint8_t col) {
    pin_t pin = col_pins[col];
    if (pin != NO_PIN) {
        setPinOutput_writeLow(pin);
        return true;
    }
    return false;
}

static void unselect_col(uint8_t col) {
    pin_t pin = col_pins[col];
    if (pin != NO_PIN) {
#            ifdef MATRIX_UNSELECT_DRIVE_HIGH
        setPinOutput_writeHigh(pin);
#            else
        setPinInputHigh_atomic(pin);
#            endif
    }
}

static void unselect_cols(void) {
    for (uint8_t x = 0; x < MATRIX_COLS; x++) {
        unselect_col(x);
    }
}


__attribute__((weak)) void matrix_init_custom(void) {
    unselect_cols();
    for (uint8_t x = 0; x < MATRIX_ROWS; x++) {
        if (row_pins[x] != NO_PIN) {
            setPinInputHigh_atomic(row_pins[x]);
        }
    }
    setPinInputHigh_atomic(B8);
}

__attribute__((weak)) void matrix_read_rows_on_col(matrix_row_t current_matrix[], uint8_t current_col, matrix_row_t row_shifter) {    // Start with a clear matrix row
        bool key_pressed = false;

    // Select col
    if (!select_col(current_col)) { // select col
        return;                     // skip NO_PIN col
    }
    matrix_output_select_delay();

    // For each row...
    for (uint8_t row_index = 0; row_index < MATRIX_ROWS; row_index++) {
        // Check row pin state
        if (readMatrixPin(row_pins[row_index]) == 0) {
            // Pin LO, set col bit
            current_matrix[row_index] |= row_shifter;
            key_pressed = true;
        } else {
            // Pin HI, clear col bit
            current_matrix[row_index] &= ~row_shifter;
        }
    }

    // Unselect col
    unselect_col(current_col);
    matrix_output_unselect_delay(current_col, key_pressed); // wait for all Row signals to go HIGH
}

bool matrix_scan_custom(matrix_row_t current_matrix[]) {
    static matrix_row_t temp_matrix[MATRIX_ROWS] = {0};

    matrix_row_t row_shifter = MATRIX_ROW_SHIFTER;
    for (uint8_t current_col = 0; current_col < MATRIX_COLS; current_col++) {
        matrix_read_rows_on_col(temp_matrix, current_col, row_shifter);
    }

    bool changed = memcmp(current_matrix, temp_matrix, sizeof(temp_matrix)) != 0;
    if (changed) {
        memcpy(current_matrix, temp_matrix, sizeof(temp_matrix));
    }
    return changed;
}
```

Reply:
I want to map it to `{"label": "CALC", "matrix": [2, 3], "x": 3, "y": 2},`

Reply:
I've added the matrix.c file in my most recent commit

Reply:
@drashna I got it working now, please ignore above.